multistream-select/src/protocol/listener.rs

```diff
@@ -178,6 +163,58 @@ where
     }
 }
 
+
+pub enum ListenerFuture<T: AsyncRead + AsyncWrite> {
+    Await {
+        inner: StreamFuture<LengthDelimitedFramedRead<Bytes, LengthDelimitedFramedWrite<T, BytesMut>>>
+    },
+    Reply {
+        sender: sink::Send<LengthDelimitedFramedRead<Bytes, LengthDelimitedFramedWrite<T, BytesMut>>>
+    },
+    Undefined
+}
+
+impl<T: AsyncRead + AsyncWrite> Future for ListenerFuture<T> {
+    type Item = Listener<T>;
+    type Error = MultistreamSelectError;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            match mem::replace(self, ListenerFuture::Undefined) {
+                ListenerFuture::Await { mut inner } => {
+                    let (msg, socket) =
+                        match inner.poll().map_err(|(e, _)| MultistreamSelectError::from(e))? {
```

Comment:
I think I prefer having `Err` in the match, since we already have a match anyway.